dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java

```diff
@@ -382,7 +382,9 @@ private String getAnalyticsSqlCacheKey( String expression, ProgramIndicator prog
     private String _getAnalyticsSql( String expression, ProgramIndicator programIndicator, Date startDate, Date endDate,
         String tableAlias )
     {
-        Set<String> uids = getDataElementAndAttributeIdentifiers( expression, programIndicator.getAnalyticsType() );
+        // Get the uids from the expression even if this is the filter
+        Set<String> uids = getDataElementAndAttributeIdentifiers( programIndicator.getExpression(),
```

Comment:
Hey @jimgrace, I'm not really familiar with this feature. So my question might not make sense, but I just want to make sure it's not a typo.

We are now using `programIndicator.getExpression()` instead of `expression`.

In the line number `400`, though -> `String sql = castString( Parser.visit( expression, visitor ) );`, we are using `expression`. Is it expected?

Reply:
Thanks @maikelarabori, good question. A program indicator has a (main) expression and a filter expression. When this code is called, `expression` will be set to one of these two. What we want to find ultimately is the value of `expression`. If we're evaluating the main expression and it contains a `V{value_count}` or `V{zero_pos_value_count}` we want to see how many of the data elements or attributes present in the main expression have a value. However if we're evaluating the filter expression, we also need to see how many of the data elements or attributes present in the main expression have a value. So to find set of data element and attribute identifiers, we should always look in the main expression, no matter whether we are evaluating the main expression or the filter expression.

Reply:
Thanks for the clarification @jimgrace :+1: